### ISSUE TITLE
EVP_PKEY_cmp() Fails Identical keys

### ISSUE THREAD

**Issue by ashman-p** [openssl/openssl#14147]
Create a EVP_PKEY and compare it to itself. EVP_PKEY_cmp() returns 0, which means a mismatch.
See code fragment below.

```
	int rv = 0;

	EVP_PKEY *mac_key = EVP_PKEY_new_raw_private_key(EVP_PKEY_HMAC, NULL,
	                                                 (const unsigned char *)("0123456789"),
	                                                 strlen("0123456789"));

    rv = EVP_PKEY_cmp(mac_key, mac_key);
    if (rv != 1) {
        printf("key compare failed. rv = %d\n", rv);
        return 0;
    }
    printf("key compare passed\n");
```

Possible fix ...
```
@ -130,8 +130,14 @@ int EVP_PKEY_cmp(const EVP_PKEY *a, const EVP_PKEY *b)
                return ret;
        }

        if (a->ameth->pub_cmp)
            return a->ameth->pub_cmp(a, b);
        if (a->ameth->pub_cmp) {
            ret = a->ameth->pub_cmp(a, b);
            if (ret == 0) {
                return 1;
            } else {
                return 0;
            }
        }
    }
```
The issue is present in the latest 1.1.1 stable branch. And appears to be on Master (3.0) as well.

**Comment by petrovr** [comment]
Norman Ashley wrote:
> [SNIP]
> Create a EVP_PKEY and compare it to itself. EVP_PKEY_cmp() returns 0, which means a mismatch.
> See code fragment below.
>
> ```
> 	int rv = 0;
>
> 	EVP_PKEY *mac_key = EVP_PKEY_new_raw_private_key(EVP_PKEY_HMAC, NULL,
> 	                                                 (const unsigned char *)("0123456789"),
> 	                                                 strlen("0123456789"));
>
>      rv = EVP_PKEY_cmp(mac_key, mac_key);
>      if (rv != 1) {
>          printf("key compare failed. rv = %d\n", rv);
>          return 0;
>      }
>      printf("key compare passed\n");
> ```

But sample is only for HMAC key while report pretend ALL keys!


> Possible fix ...
> ```
> @ -130,8 +130,14 @@ int EVP_PKEY_cmp(const EVP_PKEY *a, const EVP_PKEY *b)
>                  return ret;
>          }
>
>          if (a->ameth->pub_cmp)
>              return a->ameth->pub_cmp(a, b);
>          if (a->ameth->pub_cmp) {
>              ret = a->ameth->pub_cmp(a, b);
>              if (ret == 0) {
>                  return 1;
>              } else {
>                  return 0;
>              }
>          }
>      }
> ```
> The issue is present in the latest 1.1.1 stable branch. And appears to be on Master (3.0) as well.
>


Error is in "a-method" compares specific to a particular "key type".
Why to change a "generic" method?


Quote from EVP_PKEY_ASN1_METHOD(3):
```
The pub_cmp() method is called when two public keys are to be compared. It MUST return 1 when the keys are equal, 0 otherwise. It's called by EVP_PKEY_cmp(3) <https://www.openssl.org/docs/man1.1.1/man3/EVP_PKEY_cmp.html>.
```

Of course OpenSSL does not follow strictly documentation. Also It is long story misuse of comparison functions. EC key "cmp" is another sample in this particular case.


Roumen

**Comment by ashman-p** [comment]
Yeah, I noticed that many tests fails when i ran make test with my suggested change.

**Comment by t8m** [comment]
I cannot reproduce this with the current master. Could you please retest this?

**Comment by petrovr** [comment]
Tomáš Mráz wrote:
> I cannot reproduce this with the current master. Could you please retest this?
> 

What about 1.1.1 stable ?


Roumen

**Comment by t8m** [comment]
> What about 1.1.1 stable ? Roumen

It most probably still fails there, that's why I am keeping the branch 1.1.1 label and the issue open.

**Comment by petrovr** [comment]
Румен Петров wrote:
> Tomáš Mráz wrote:
>> I cannot reproduce this with the current master. Could you please retest this?
>>
> 
> What about 1.1.1 stable ?
> 
> 
> Roumen
> 
> 


Quote from https://www.openssl.org/docs/man1.1.0/man3/EVP_PKEY_ASN1_METHOD.html
---
The pub_cmp() method is called when two public keys are to be compared. It MUST return 1 when the keys are equal, 0 otherwise. It's called by EVP_PKEY_cmp(3).
---
Then see (1) hmac/hm_ameth.c and (2) ec/ec_ameth.c both pub_cmp does not follow documentation. The code in (2) is harmless in context of EVP_PKEY_cmp but in (1) is used ASN1_OCTET_STRING_cmp
Outside EVP_PKEY_cmp (2) must be considered as defect.


Quote from https://www.openssl.org/docs/man1.1.0/man3/ASN1_STRING_cmp.html
---
ASN1_STRING_cmp() compares a and b returning 0 if the two are identical. The string types and content are compared.
---

Roumen

**Comment by t8m** [comment]
I see both on master and 1.1.1 that ec_ameth ec_key_pub_cmp returns 0 if EC_POINT_cmp() returns nonzero and 1 EC_POINT_cmp() returns 0. That is IMO correct.

The hmac implementation is broken. I'll submit PR for 1.1.1 (it is not there anymore in master at all).

**Comment by petrovr** [comment]
Tomáš Mráz wrote:
> I see both on master and 1.1.1 that ec_ameth ec_key_pub_cmp returns 0 if EC_POINT_cmp() returns nonzero and 1 EC_POINT_cmp() returns 0. That is IMO correct.
> 
Sure but let see code:
---
static int eckey_pub_cmp(const EVP_PKEY *a, const EVP_PKEY *b)
{
     int r;
     const EC_GROUP *group = EC_KEY_get0_group(b->pkey.ec);
     const EC_POINT *pa = EC_KEY_get0_public_key(a->pkey.ec),
         *pb = EC_KEY_get0_public_key(b->pkey.ec);
     if (group == NULL || pa == NULL || pb == NULL)
         return -2;
     r = EC_POINT_cmp(group, pa, pb, NULL);
     if (r == 0)
         return 1;
     if (r == 1)
         return 0;
     return -2; /* <==== non allowed return value for EVP_PKEY_ASN1_METHOD pub_cmp() method */
}
---
Reviewing the code is visible that method could return -2.
At this point if EC_POINT_cmp returns values different from 0 and 1 what will be result?

Note "EC_POINT_cmp returns 1 if the points are not equal, 0 if they are, or -1 on error".

So if "EC_POINT_cmp fails method return -2. This is not allowed for pub_cmp() - see https://www.openssl.org/docs/man1.1.1/man3/EVP_PKEY_asn1_set_param.html



> The hmac implementation is broken. I'll submit PR for 1.1.1 (it is not there anymore in master at all).
> 


Roumen

**Comment by t8m** [comment]
I do not think it is relevant as EVP_PKEY_cmp is documented to have negative return values on errors.

**Comment by t8m** [comment]
And the pub_cmp() is not used outside of that call. And yeah, in theory you could obtain the ameth and obtain the pub_cmp() from it and call it directly from an application. Unless you have an example of such application I am not going to spend more time on that theoretical concern.

**Comment by t8m** [comment]
Fixed by ccfe5ec
